### PR TITLE
Binary variadic introduces null

### PR DESCRIPTION
**Pull request by mgree**
### Motivation

This PR addresses #13929, making the `introduces_nulls` predicates explicit and complete for `BinaryFunc` and `VariadicFunc`.

I couldn't get tests to run locally (ancient laptop), so letting CI do its thing.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by ggevay — CHANGES_REQUESTED** [review]
Thanks! I have a few comments.

Could you please add a test whose output changes due to your improvement? Currently, `introduces_nulls` is only used in `decompose_is_null`, so you will need a query that `decompose_is_null` changes. (Later, it will be used for more optimizations, see https://github.com/MaterializeInc/materialize/issues/17940) `not-null-propagation.slt` would be a good place for the test.

Also, I'm wondering whether we should try to add a similar test to `unary_func_introduces_nulls`. But the MzReflect trait [might be deprecated soon](https://github.com/MaterializeInc/materialize/issues/17917), so maybe not? What do you think @aalexandrov?

**Comment by ggevay** [comment]
Btw. you will need to be added to the "code-authorized" team to run CI. I've hit Rebuild on Buildkite to run it with my credentials in the meantime.

**Comment by ggevay** [comment]
@antiguru, could you please add @mgree [to the code authorized team](https://www.notion.so/materialize/Engineering-overview-868ccd6c632349d19b095b4825bd42f8?pvs=4#b228e3e18d7248a8b47611f3ea9b7a78)?

File: src/expr/src/scalar/func.rs
Comment (ggevay) [inline src/expr/src/scalar/func.rs:6741]: When does `replace` introduce nulls?

File: src/expr/src/scalar/func.rs
Comment (ggevay) [inline src/expr/src/scalar/func.rs:6743]: It seems to me that `JsonbBuildObject` and `JsonbBuildArray` return null only if one of the inputs are null.

**Comment by antiguru** [comment]
> @antiguru, could you please add @mgree [to the code authorized team](https://www.notion.so/materialize/Engineering-overview-868ccd6c632349d19b095b4825bd42f8?pvs=4#b228e3e18d7248a8b47611f3ea9b7a78)?

That will be fixed next week or so!

File: src/expr/src/scalar/func.rs
Comment (ggevay) [inline src/expr/src/scalar/func.rs:6744]: `ArrayToString` should be false, unless I'm misunderstanding something. Here are the docs: https://materialize.com/docs/sql/functions/#array_to_string

File: src/expr/src/scalar/func.rs
Comment (ggevay) [inline src/expr/src/scalar/func.rs:6747]: There might be a similar issue as with `IsRegexpMatch` above.

File: src/expr/src/scalar/func.rs
Comment (mgree) [reply src/expr/src/scalar/func.rs:6744]: You're right---thrown off by manual handling.

File: src/expr/src/scalar/func.rs
Comment (mgree) [reply src/expr/src/scalar/func.rs:6741]: Good catch!

File: src/expr/src/scalar/func.rs
Comment (mgree) [reply src/expr/src/scalar/func.rs:6743]: You're right; I misinterpreted the slice of `Datum`s.

File: src/expr/src/scalar/func.rs
Comment (mgree) [reply src/expr/src/scalar/func.rs:6747]: `RegexpMatch` is implemented by `regexp_match_dynamic` which defers to `regexp_match_static` which returns null when there are no valid captures/no match. I don't 100% understand the way row packing works... if it packs just one null, will that be interpreted as a singleton array of nulls or a null proper?

File: src/expr/src/scalar/func.rs
Comment (mgree) [reply src/expr/src/scalar/func.rs:2672]: You're right---it doesn't.

Actual behavior in `is_like_match_dynamic` and `is_regexp_match_dynamic` is to error on invalid pattern.

I don't know where to look in the docs beyond https://materialize.com/docs/sql/functions/, which document the existence of the `~` opeartor but not its behavior. `LIKE` is similarly not documented in a detailed way (there, at least).

In Postgres, invalid like patterns (`select 'abc' like '\';`) error out; regexes behave similarly (`select 'abc' similar to '*';`).

Should I update `BinaryFunc::output_type` in a fix?

File: src/expr/src/scalar/func.rs
Comment (ggevay) [reply src/expr/src/scalar/func.rs:6746]: Ah, yes! It can indeed return null if there is no match. The `None => packer.push(Datum::Null),` does this, and there is no array created in this case. (The `Some` case uses `push_array` to construct an array (with possibly nulls inside from the inner `None` case).)

File: src/expr/src/scalar/func.rs
Comment (ggevay) [reply src/expr/src/scalar/func.rs:2672]: Ok, then I'd say
- The behavior is fine, since it matches Postgres (and makes sense, which is also nice :)).
- `IsRegexpMatch` and `IsLikeMatch` should both be `false`.
- The docs are fine. I'd say it's not necessary to document that we error out on an invalid input. I was just worried about the docs possibly incorrectly saying that we output null. (Btw. if the behavior would be outputting null, that would have needed documentation IMO.)
- Yes, please fix `BinaryFunc::output_type`. This can go in this same PR.

File: src/expr/src/scalar/func.rs
Comment (mgree) [reply src/expr/src/scalar/func.rs:6746]: Makes sense... thanks!

File: src/expr/src/scalar/func.rs
Comment (mgree) [reply src/expr/src/scalar/func.rs]: Haha, always nice to have things make sense! 😁

**Comment by ggevay** [comment]
Now there is a test failure in `not-null-propagation.slt`, probably because of the `typ` change.

**Comment by mgree** [comment]
I'll take a look at this failing test and write a new test this week---I haven't looked at/been shown the testing infrastructure in any detail yet (main laptop is too old to run the version of Docker that can run the test suite), so it might take a me a minute to get this bit ironed out.

**Comment by mgree** [comment]
> Could you please add a test whose output changes due to your improvement? Currently, `introduces_nulls` is only used in `decompose_is_null`, so you will need a query that `decompose_is_null` changes. (Later, it will be used for more optimizations, see #17940) `not-null-propagation.slt` would be a good place for the test.

Do the added tests on variadic functions suffice, or should I write more?

NB I added a test in table_func.slt as well, per some confusing behavior in the jsonb functions #18059.

**Comment by aalexandrov** [comment]
> Do the added tests on variadic functions suffice, or should I write more?

Do we have tests for the binary function changes that you did already? If not, maybe add one more test to `non-null-propagation.slt`.

**Comment by mgree** [comment]
There were some, but only one changed: https://github.com/MaterializeInc/materialize/blob/0d6eed5f239ccd351786237a69660eb2bf15517c/test/sqllogictest/not-null-propagation.slt#L297-L307

Happy to knock out a few more on Wednesday morning. Should I pick a few interesting things that wouldn't have passed before, or just go for exhaustiveness?

File: src/expr/src/scalar/func.rs
Comment (aalexandrov) [inline src/expr/src/scalar/func.rs:6680]: I'm not sure how `error_if_null` can ever produce a `null` output (even if the input is `null`). Can you give a counter-example?

**Review by aalexandrov — APPROVED** [review]
I have only once specific question.

If we don't include extensive unit or SLT tests for every function, let's ask @MaterializeInc/qa to run some regression tests to catch for potential regressions.

In general, my worry is that without investment in dedicated unit tests the chance of implementation and annotation methods running out of sync is higher.

**Comment by ggevay** [comment]
> If we don't include extensive unit or SLT tests for every function, let's ask @MaterializeInc/qa to run some regression tests to catch for potential regressions.

Does Philip's RQG testing include calls to various exotic scalar functions? If not, then I guess testing this wouldn't be straightforward at all for @MaterializeInc/qa.

Instead, we could add a unit test similar to `unary_func_introduces_nulls` for binary and variadic functions. (The use of `mz-reflect` [could be avoided](https://github.com/MaterializeInc/materialize/issues/17917#issuecomment-1456000840) by just explicitly including a list of functions.) @mgree, how do you feel about this? If you'd rather move on, e.g. to https://github.com/MaterializeInc/materialize/issues/17909, then we can open a separate issue to do this later, and just merge this PR soon.

There are a few minor things that I would like to check on the PR before merging, which I hope to do tomorrow.

File: src/expr/src/scalar/func.rs
Comment (mgree) [reply src/expr/src/scalar/func.rs]: Nope---you're right!

**Comment by mgree** [comment]
I'm happy to spend a bit of time writing exhaustive tests on Wednesday morning, but I won't be mad if someone else does it before then. 😁

**Comment by aalexandrov** [comment]
> I'm happy to spend a bit of time writing exhaustive tests on Wednesday morning, but I won't be mad if someone else does it before then.

Just do as much as you want and push here before you move on to the next issue. I'll pick it up and add the remaining tests before I merge this on Friday.

**Comment by aalexandrov** [comment]
FYI the reason why we want these tests is because a false negative in `introduces_nulls` potentially might lead to incorrect transformations.

I now realize that demonstrating that
> For all inputs that are not null function x will always produce something that is not null

is somehow much harder than demonstrating that
> For some inputs that are not null function x will produce something that null

Based on that, doing an extensive test coverage only for the `introduces_nulls = true` case seems a bit of an overkill. I would still want to go over the `eval` definitions to validate the `introduces_nulls = false` case before I merge this, and this most likely will happen on Friday.

**Comment by mgree** [comment]
I agree---the negated existential (i.e., universal) quantifier is a bear. I think going over my work is definitely a good idea, considering you and Gábor have both found bugs just by inspection. 😅

I'll write some of these tests to warm up tomorrow morning; I'll leave a comment after I push to make it clear I'm moving on.

Thanks for your patient help on this! 😁

**Comment by ggevay** [comment]
> it's not obvious to me that COALESCE, GREATEST, and LEAST are introducing nulls---if non of the inputs have nullable types, then the only way to get a null out is to have an empty argument list (which is banned by the parser)...

Hmm, great observation! I'd say let's make `introduces_nulls` and `output_type` precise for at least COALESCE, because COALESCE is quite a common way to make something non-null.

For COALESCE (but _not_ for GREATEST and LEAST), the `output_type` should be non-nullable when at least one of the arguments is non-nullable. Note that this is the _opposite_ of what the `union` in the `reduce` is doing. (Btw. I'm wondering if all that other stuff that `union` is doing are actually correct for COALESCE...) This would be great in case the user writes something like `COALESCE(my_nullable_int_column, 0)`. We could add a test like this:
```
CREATE TABLE t2 (k1 INTEGER, k2 INTEGER not null);
EXPLAIN SELECT * FROM t2 WHERE COALESCE(k1,k2) IS NULL;
```
which should optimize to `Constant <empty>` if `output_type` is upgraded as per above.

However, we should be careful when relying on the non-empty argument list invariant: [I asked it in the team-surfaces channel whether we can rely on this behavior.](https://materializeinc.slack.com/archives/C02FWJ94HME/p1678994188182589) Let's wait for the (hopefully yes) answer there. And then if we go ahead, then still I'd say we should proactively guard ourselves against a situation where somebody removes this invariant from the parser, or makes the optimizer pipeline introduce a COALESCE/GREATEST/LEAST with an empty argument list in some other way. First, note that the `unwrap` call after the `reduce` in `output_type` would explode for an empty argument list. At that line, we should also add a comment with a warning that if somebody changes this behavior then attend to both `introduces_nulls` and the nullability of the `output_type`. Additionally:
- The `coalesce`/`greatest`/`least` implementation functions should have an `assert` that the arg list is not empty. This line should have the same warning comment as above.
- It would be good to create slt tests that call COALESCE, LEAST, GREATEST with an empty arg list and expect the parser error. (An slt test will expect an error if the line starts with `query error`, and then the error msg.) The comment above the test should have the same warning.

(Or if you would rather move on from this issue, then we can open a separate issue for this, and tackle it later.)

**Comment by ggevay** [comment]
Or wait a moment: https://github.com/MaterializeInc/materialize/blob/4fd03051120c51808ce01496885af57581004deb/src/expr/src/scalar/mod.rs#L1085 _will_ make a COALESCE with an empty arg list sometimes... And then what would happen is that https://github.com/MaterializeInc/materialize/blob/4fd03051120c51808ce01496885af57581004deb/src/expr/src/scalar/mod.rs#L1111 would explode if we have the assert in the implementation function that it is non-empty.

We could move up the `if exprs.iter().all(|e| e.is_literal()) {` branch to before the `*func == VariadicFunc::Coalesce` branch. But this is getting a bit dangerous, so maybe just open an issue and postpone.

File: test/sqllogictest/not-null-propagation.slt
Comment (ggevay) [inline test/sqllogictest/not-null-propagation.slt:330]: should be Bs instead of Is

File: test/sqllogictest/not-null-propagation.slt
Comment (ggevay) [inline test/sqllogictest/not-null-propagation.slt:338]: same here

**Review by ggevay — APPROVED** [review]
Looks good, apart from one last minor comment.

And for the COALESCE/GREATEST/LEAST stuff, I'd suggest to optionally address that in a follow-up PR.

**Comment by ggevay** [comment]
(Please squash all commits into 1 commit before merging.)

**Pull request merged**
